docs/design/sql/02-operator-interface.md

```diff
@@ -0,0 +1,157 @@
+structure of the parse tree.
+
+A **query plan**, consisting of a relational tree and supplemental information, is submitted for execution
+after the optimization.
+
+The table below lists common relational operators used in database engines.
+
+*Table 1: Common Relational Operators*
+
+| Name | Description |
+|---|---|
+| `Scan` | Iterate over source rows |
+| `Project` | Return a set of original or derived attributes of the child operator |
+| `Filter` | Return rows of the child operator which pass the provided predicate |
+| `Aggregate` | Aggregate rows of the child operator |
+| `Sort` | Sort rows of the child operator |
+| `Join` | Join rows from several child operators |
+
+An example of a query, its parse tree, and its relational tree is provided below.
+
+*Snippet 1: Query*
+```sql
+SELECT a, SUM(b)
+FROM table
+GROUP BY a
+HAVING SUM(b) > 50
+```
+*Snippet 2: Parse Tree*
+```
+-- Select
+---- SelectList [a, SUM(b)]
+---- From [table]
+---- GroupBy [a]
+---- Having [SUM(b) > 50]
+```
+*Snippet 3: Relational Tree*
+```
+-- Filter [SUM(b) > 50]
+---- Aggregate [a -> SUM(b)]
+------ Project [a, b]
+-------- Scan [table]
+```
+
+## Volcano Model
+
+Volcano Model defines the common data exchange interface between operators in the relational tree. This allows
+for extensibility, as new operators could be implemented with minimal changes to the engine.
+
+In the original paper the interface consists of three operations:
+
+*Snippet 4: Volcano Interface*
+```java
+interface Operator {
+    void open();  // Initialize the operator
+    Row next();   // Get the next row
+    void close(); // Close the operator and release all resources
+}
+```
+
+## Volcano Model in Hazelcast Mustang
+
+The original Volcano Model has two drawbacks:
+1. Operators exchange one row at a time, which leads to performance overhead
+2. Call to the `next()` is blocking, which is not optimal for the distributed environment, where
+operators often wait for remote data or free space in the send buffer.
+
+To achieve high performance, we introduce several changes to the original Volcano Model: batching and
+non-blocking execution.
+
+### Row and RowBatch
+We define the `RowBatch` interface which a collection of rows (tuples).
+
+*Snippet 5: RowBatch interface*
+```java
+interface RowBatch {
+    Row getRow(int index); // Get the row by index
+    int getRowCount();     // Get the number of rows 
+} 
+```
+
+Then we define the `Row` interface, which provides access to values by index. The `Row` itself is considered
+as a special case of `RowBatch` with one row. This allows saving on allocations in some parts of the engine.
+
+*Snippet 6: Row interface*
+```java
+interface Row extends RowBatch {
+    Object get(int index); // Get the value by index
+    int getColumnCount();  // Get the number of values in the row 
+    
+    default int getRowCount() {
+        return 1;
+    }
+    
+    default int getRow(int index) {
```

Comment:
The return type should be `Row `, I believe :)

Reply:
Fixed.